File: ruido_correlacionado.py
# coding=utf-8
from numpy import * 
from scipy import * 
from matplotlib.pyplot import *

# Geração de ruído correlado: 
# X => vector gaussiano com N = 20  
# matriz de covariância: 
# E[Xi Xj] = e^|i-j|/5, i,j = 1,...,20. 
# Decomposição de Cholesky => Y = B X

N = 20
I = range(1,N+1)
n = 1000
X = random.randn(N,n)
Y = zeros((N,n))
C = zeros((N,N))

for i in range(N):
    for j in range(N):
        C[i,j] = exp(-abs(i-j)/5)

B = linalg.cholesky(C).T

for i in range(n):
    Y[:,i] = B.dot( X[:,i] )

Ce = cov( Y )

figure(1)
imshow(C, cmap='Greys',extent=[1,N,1,N],
aspect='auto',origin='lower')
xlabel(r'$i$',fontsize=18)
ylabel(r'$j$',fontsize=18)

figure(2)
imshow(Ce, cmap='Greys',extent=[1,N,1,N],
aspect='auto',origin='lower')
xlabel(r'$i$',fontsize=18)
ylabel(r'$j$',fontsize=18)

show()

File: ruido_correlacionado.m
% Fundamentos de Acústica Submarina 

% Geração de ruído correlado: 
% X => vector gaussiano com N = 20  
% matriz de covariância: 
% E[Xi Xj] = e^|i-j|/5, i,j = 1,...,20. 
% Decomposição de Cholesky => Y = B X

clear all, close all 

N = 20;
n = 1000;
X = randn(N,n);
Y = zeros(N,n);
C = zeros(N,N);

for i = 1:N
    for j = 1:N
        C(i,j) = exp(-abs(i-j)/5);
    endfor
endfor

B = chol(C);

Y = B*X;

Ce = cov( Y' );

colormap('jet')

figure(1)
pcolor(C), shading interp, colorbar
xlabel('i','FontSize',18)
ylabel('j','FontSize',18)

figure(2)
pcolor(Ce), shading interp, colorbar
xlabel('i','FontSize',18)
ylabel('j','FontSize',18)

